### PR TITLE
✨ Update MHC proposal to use conditions instead of annotations

### PR DESCRIPTION
**Pull request by benmoss**
**What this PR does / why we need it**:
We realized during the review of #2975 that using annotations would amount to a privilege escalation for users who had `machine.edit` access but not `machine.delete`.

So instead we're proposing that we use conditions (#3017) because that has the benefit of being a part of the status, which requires a higher level of privilege that is not commonly held by users.

That CAEP is still pending so this is effectively blocked on that proposal being merged first.

/hold

xref #2836, #2920

### CODE REVIEW

File: docs/proposals/20191030-machine-health-checking.md [hunk 1]
```diff
@@ -105,14 +104,12 @@ As an operator of a Management Cluster, I want my machines to be self-healing an
 
 ### Implementation Details/Notes/Constraints
 
-#### Machine annotation:
+#### Machine condition:
 ```go
-const MachineUnhealthyAnnotation = "machine.cluster.x-k8s.io/unhealthy"
+const ConditionHealthCheckPassed ConditionType = "HealthCheckPassed"
```
Comment (vincepri) [inline docs/proposals/20191030-machine-health-checking.md:109]:
```suggestion
const ConditionHealthCheckSucceeded ConditionType = "HealthCheckSucceeded"
```

IIRC we're trying to standardize on Ready/Succeeded for the first iteration
cc @fabriziopandini

File: docs/proposals/20191030-machine-health-checking.md [hunk 2]
```diff
@@ -105,14 +104,12 @@ As an operator of a Management Cluster, I want my machines to be self-healing an
 
 ### Implementation Details/Notes/Constraints
 
-#### Machine annotation:
+#### Machine condition:
```
Comment (vincepri) [inline docs/proposals/20191030-machine-health-checking.md:107]: Who owns this condition? Machine or MachineHealthCheck?

Ownership is probably on the Machine, while operations are done outside?

File: docs/proposals/20191030-machine-health-checking.md [hunk 3]
```diff
@@ -167,7 +164,7 @@ type target struct {
 - Annotates unhealthy target machines as described above.
 
 Out of band:
-- The owning controller observes the annotation and is responsible to remediate the machine.
+- The owning controller observes the condition and is responsible to remediate the machine.
```
Comment (vincepri) [inline docs/proposals/20191030-machine-health-checking.md:167]: This shouldn't be true anymore, it should be a different condition to signal remediation

File: docs/proposals/20191030-machine-health-checking.md [hunk 4]
```diff
@@ -177,13 +174,32 @@ Out of band:
...
+If a user configures healthchecks such that more than one check applies to
+machines these healthchecks can potentially become a source of contradictory
+information. For instance, one healthcheck could pass for a given machine,
+while another fails. This will result in undefined behavior since the order in
+which the checks are run and the timing of the remediating controller will
+determine which condition is used for remediation decisions.
+
+There are safeguards that can be put in place to prevent this but it is out of
+scope for this proposal. Documentation should be updated to warn users to
+ensure their healthchecks do not overlap.
```
Comment (vincepri) [inline docs/proposals/20191030-machine-health-checking.md:188]: Potentially, `clusterctl` could be extended to do validation/conformance in the future

File: docs/proposals/20191030-machine-health-checking.md (hunk 3 above)
Comment (vincepri) [reply docs/proposals/20191030-machine-health-checking.md:167]: We discussed about having `RemediationByOwnerSucceeded` or `OwnerRemediationSucceded` for example which is a contract that a Machine defines (and own), but other controllers have to respect.

File: docs/proposals/20191030-machine-health-checking.md (hunk 2 above)
Comment (benmoss) [reply docs/proposals/20191030-machine-health-checking.md:107]: What do you mean by ownership?

File: docs/proposals/20191030-machine-health-checking.md (hunk 3 above)
Comment (benmoss) [reply docs/proposals/20191030-machine-health-checking.md:167]: Can you explain why we need more than one condition?

File: docs/proposals/20191030-machine-health-checking.md (hunk 4 above)
Comment (benmoss) [reply docs/proposals/20191030-machine-health-checking.md:188]: That seems weak since it's not a mandatory entrypoint to the system. Validations in clusterctl should only be a better UX for validations that happen in the actual system, similar issue as #3049

File: docs/proposals/20191030-machine-health-checking.md (hunk 2 above)
Comment (vincepri) [reply docs/proposals/20191030-machine-health-checking.md:107]: Ownership is usually given to the object/controller that defines the contract

File: docs/proposals/20191030-machine-health-checking.md (hunk 3 above)
Comment (vincepri) [reply docs/proposals/20191030-machine-health-checking.md:167]: Machine type should define the healthy annotation to determine if it's healthy or not, but any sort of remediation should probably go in a different condition, specific to which strategy should be taken.

For example:
- `OwnerRemediationSucceded` is applied when remediation is needed and we expect the controller-owner to remediate the Machine.
- `ExternalRemediationSucceeded` is instead applied when remediation is expected to happen outside, and the owning controllers shouldn't take any actions to remediate.

These would be defined in the Machine type, but the contract would state that they would be applied by a health check controller (MHC in this case).

cc @JoelSpeed (pls ping others from External Remediation as well)

Comment (JoelSpeed) [reply docs/proposals/20191030-machine-health-checking.md:167]: cc @beekhof @n1r1

Comment (benmoss) [reply docs/proposals/20191030-machine-health-checking.md:167]: Can we punt the details of external remediation to that proposal? I don't see why we need to resolve that here

Comment (vincepri) [reply docs/proposals/20191030-machine-health-checking.md:167]: We can, but we should at least specify OwnerRemediationSucceded

File: docs/proposals/20191030-machine-health-checking.md [hunk 5]
```diff
@@ -105,14 +104,12 @@ As an operator of a Management Cluster, I want my machines to be self-healing an
 
 ### Implementation Details/Notes/Constraints
 
-#### Machine annotation:
+#### Machine condition:
 ```go
-const MachineUnhealthyAnnotation = "machine.cluster.x-k8s.io/unhealthy"
+const ConditionHealthCheckSucceeded ConditionType = "HealthCheckSucceeded"
```
Comment (fabriziopandini) [inline docs/proposals/20191030-machine-health-checking.md:109]: I'm definitely not good at naming, but since this condition represent an ongoing status, this should be end in `Ready`. What about “HealthzReady“?

File: docs/proposals/20191030-machine-health-checking.md [hunk 6]
```diff
@@ -167,7 +164,7 @@ type target struct {
 - Annotates unhealthy target machines as described above.
 
 Out of band:
-- The owning controller observes the annotation and is responsible to remediate the machine.
+- The owning controller observes the condition and is responsible to remediate the machine.
 - The owning controller performs any pre-deletion tasks required.
```
Comment (n1r1) [inline docs/proposals/20191030-machine-health-checking.md:168]: What is expected from the owning controller if the Machine becomes healthy again during/before some of the "pre-deletion" tasks?

What do we expect from MHC to do if it marked the Machine as unhealthy with that condition and later it becomes healthy again?

File: docs/proposals/20191030-machine-health-checking.md (hunk 3 above)
Comment (JoelSpeed) [reply docs/proposals/20191030-machine-health-checking.md:167]: This has a lot of overlap with the external remediation proposal. For example, if this condition approach is a suitable signalling mechanism, then the external remediation CRDs (or at least MHC creating them from templates) doesn't add that much value

Why could KCP not use the external remediation proposal to create a `KCPMachineRemediation` CRD to act as the signal for requiring remediation?

This is kind of an aside, but I'm not completely familiar with the actions that KCP needs to perform before deleting the machine and I'd like to understand this a bit better, does anyone know if this is documented somewhere?

File: docs/proposals/20191030-machine-health-checking.md (hunk 4 above)
Comment (JoelSpeed) [reply docs/proposals/20191030-machine-health-checking.md:188]: This validation would be better in a webhook, though I'm not sure how it would work, we can try to validate that MHC selectors don't overlap, but that's complex and not guaranteed to work. So then we have to check that MachineSets don't create machines with labels that overlap in MHC.

This relates back to having multiple remediations at play at once that are independent and don't know about one another. Perhaps we should revisit the hooks proposal as this would prevent this kind of issue I believe

File: docs/proposals/20191030-machine-health-checking.md (hunk 6 above)
Comment (benmoss) [reply docs/proposals/20191030-machine-health-checking.md:168]: This is the same as with annotations, machines do not become healthy again right now, they just are deleted. If it passes a MHC after it fails, it will still get deleted.

File: docs/proposals/20191030-machine-health-checking.md (hunk 5 above)
Comment (benmoss) [reply docs/proposals/20191030-machine-health-checking.md:109]: Hmm, I'm not sure I get what you mean. To me this condition represents that the last health check we ran succeeded. Nothing is "ready" as a result.

File: docs/proposals/20191030-machine-health-checking.md (hunk 3 above)
Comment (benmoss) [reply docs/proposals/20191030-machine-health-checking.md:167]: Yeah, that's why I'm trying to push that discussion out of this issue. This is just an update to the annotation proposal that uses conditions rather than annotations.

Comment (JoelSpeed) [reply docs/proposals/20191030-machine-health-checking.md:167]: > We can discuss separately whether conditions vs CRDs make sense for external remediation.

Personally I think we should be having that discussion before we do the KCP remediation work.

The concern I have is that MHC is becoming complicated with lots of different paths to various remediations that might trample one another

Ie, if it's a machine it gets an annotation, unless theres an object reference to a template, but if it's a KCP machine, it gets an annotation even if there is an object reference to a template? 

I can see this becoming very confusing and would prefer that we have a discussion about how we want this to look long term. I don't mind baby steps to get there, I'm in favour of that, but at the moment, I don't think we have a long term goal or vision and I'm concerned that we might cause more problems if we implement stuff without taking a step back and looking at it.

I don't want to block the effort of getting KCP remediation working, I can see conditions working as an intermediary, but I want to have the conversation about aligning, long term, either external needs to be triggered by conditions too, or KCP needs to be CRD based, or they both need to change

File: docs/proposals/20191030-machine-health-checking.md (hunk 6 above)
Comment (n1r1) [reply docs/proposals/20191030-machine-health-checking.md:172]: I understand that this is the goal but I'm curios about the implementation details.
What is expected from MHC if unhealthy Machine is healthy again? should it change the status of `ConditionHealthCheckSucceeded` from `False` to `True`?

Comment (JoelSpeed) [reply docs/proposals/20191030-machine-health-checking.md:172]: I think the intention is that the MHC should never change this condition, once it's been added and set to false, that's it, that Machine is going to go away, there's no point health checking it anymore.

I think once external remediation is implemented, it will need to be reconsidered, but for the purposes of this proposal, having it "irreversible" is sufficient

File: docs/proposals/20191030-machine-health-checking.md [hunk 7]
```diff
@@ -164,10 +165,10 @@ type target struct {
 
 - Calculate the number of unhealthy targets.
 - Compare current number against `maxUnhealthy` threshold and temporary short circuits remediation if the threshold is met.
-- Annotates unhealthy target machines as described above.
+- Marks unhealthy target machines with the conditions as described above.
 
 Out of band:
-- The owning controller observes the annotation and is responsible to remediate the machine.
+- The owning controller observes the OwnerRemediationSucceeded=False condition and is responsible to remediate the machine.
 - The owning controller performs any pre-deletion tasks required.
```
Comment (enxebre) [inline docs/proposals/20191030-machine-health-checking.md:172]: I know is not introduced by this PR but could we elaborate on what are `pre-deletion tasks`?

To make sure we don't miss any gap and illustrate better how and when the conditions are expected to interact, could we please flesh out here the details and  workflow for each of the remediation processes owned by capi here, i.e regular vs controlPlane machines?

File: docs/proposals/20191030-machine-health-checking.md [hunk 8]
```diff
@@ -177,13 +178,32 @@ Out of band:
...
+## Contradictory signal
+
+If a user configures healthchecks such that more than one check applies to
+machines these healthchecks can potentially become a source of contradictory
+information. For instance, one healthcheck could pass for a given machine,
+while another fails. This will result in undefined behavior since the order in
+which the checks are run and the timing of the remediating controller will
+determine which condition is used for remediation decisions.
+
+There are safeguards that can be put in place to prevent this but it is out of
```
Comment (enxebre) [inline docs/proposals/20191030-machine-health-checking.md:190]: >There are safeguards that can be put in place to prevent this but it is out of scope for this proposal

At the moment although we could find ways to improve UX there's a predictable and known behaviour.  A single Machine is liable to be remediated and so deleted by any MHC.

This PR is breaking this behaviour and resulting in an unknown behaviour. I believe solving this needs to be in scope.

Comment (enxebre) [reply docs/proposals/20191030-machine-health-checking.md:190]: I strongly support integrating with KCP. I think to get us there we need to have an overall detailed plan which results at minimum on the same UX we have today but we should not degrade it.

File: docs/proposals/20191030-machine-health-checking.md (hunk 6 above)
Comment (n1r1) [reply docs/proposals/20191030-machine-health-checking.md:172]: Makes sense. I think this should be explicitly specified in the proposal

File: docs/proposals/20191030-machine-health-checking.md (hunk 7 above)
Comment (vincepri) [reply docs/proposals/20191030-machine-health-checking.md:172]: We should probably remove that goal, it's not relevant or in scope for this proposal

**Comment by vincepri** [comment]
@n1r1 @enxebre @JoelSpeed Lots of conversations happening above, I'll try to summarize here what we're trying to achieve and why.

#### Goals
- Migrate the logic from `annotations` to `status.conditions`.
- Add a `HealthCheckSucceeded` condition to Machine objects.
  - MachineHealthCheck controller watches Machines, and if it doesn't find the Condition, writes one.
  - MachineHealthCheck during subsequent checks, if one fails it transitions the condition's `State` to `False`.
    > Note: for now we'll only do this once, in the future we can chat about allowing to being flipped back, but probably only after remediation — this discussion should happen in the external proposal)
- Add a `OwnerRemediationSucceded` condition to Machine objects.
  - MachineHealthCheck controller, when it sees that a Machine goes unhealthy, it writes also this condition with its `State` set to `False`.
  - Owner-Controllers watching these Machines (i.e. KCP, or MachineSet) trigger their own built-in remediation based on `OwnerRemediationSucceded` being False.
  - Owner-Controllers will, in this case, try to ultimately delete the Machine and set `OwnerRemediationSucceded` to `True` after setting the deletion timestamp.
    > Note: if the patch to set the condition state after deleting the Machine fails, we'll fail gracefully given that is expected. In most cases, we'll be able to patch the condition given that deleting a Machine isn't a quick operation (except for CAPD probably).


#### Observations
- Adding two conditions instead of a single one allows us to expand (in the future) the remediations we want to support (like external) by either not writing the second condition, or writing a different one based on some strategy (details tbd in the other proposal).
- In the future, I can see users potentially asking to disable remediation completely (e.g. by setting a `remediation.strategy = none`) — this use case is going to be fully covered by allowing MHC to not write any remediation condition.

File: docs/proposals/20191030-machine-health-checking.md (hunk 6 above)
Comment (benmoss) [reply docs/proposals/20191030-machine-health-checking.md:172]: I actually think we should update it to true. There's always a chance that the controller has not done remediation yet, so in the rare event that it fixes itself and hasn't been remediated yet we should leave it alone

File: docs/proposals/20191030-machine-health-checking.md (hunk 7 above)
Comment (benmoss) [reply docs/proposals/20191030-machine-health-checking.md:173]: For the sake of the contract all that matters is that the owning controller handles the conditions. I don't even think it's a requirement anymore that we delete the machine, since this proposal introduces a mechanism for marking it as successfully remediated.

File: docs/proposals/20191030-machine-health-checking.md (hunk 6 above)
Comment (detiber) [reply docs/proposals/20191030-machine-health-checking.md:173]: What if a controller has only performance partial remediation (prior to the deletion of the Machine resource) at the point that the Machine is returned to healthy? Do we need to worry about the possibility of the Machine being treated as healthy when it may actually now be in a degraded state from the partial remediation?

In a simplistic case, what if we have already cordoned/drained a given Node associated with a Machine but have not yet performed the deletion when the MHC now marks the Machine as ready, if for some reason the cordoning/draining and machine deletion are handled through re-entrancy rather than atomically, we could end up with a situation where we've cordoned/drained the Node, but fail to delete the Machine and now that the Machine is "healthy" we fail to delete the Machine, but also have a Machine that is also not capable of handling workloads due to being cordoned.

File: docs/proposals/20191030-machine-health-checking.md [hunk 9]
```diff
@@ -105,14 +104,17 @@ As an operator of a Management Cluster, I want my machines to be self-healing an
...
 ```go
-const MachineUnhealthyAnnotation = "machine.cluster.x-k8s.io/unhealthy"
+const ConditionHealthCheckSucceeded ConditionType = "HealthCheckSucceeded"
+const ConditionOwnerRemediationSucceeded ConditionType = "OwnerRemediationSucceeded"
 ```
 
-This annotation is applied by the MHC controller and it is expected that it
-will never be removed. The only current mediation strategy is deletion, which
-is to be handled by the machine's owning controller.
+Both of these conditions are applied by the MHC controller. ConditionHealthCheckSucceeded should only be updated by the MHC after running the health check.
```
[the remark this answers is not in the document]
Comment (benmoss) [reply docs/proposals/20191030-machine-health-checking.md:113]: I take it this is related to [your comment](https://github.com/kubernetes-sigs/cluster-api/pull/3017#discussion_r426704167) about relaxing the need for truthy conditions always being "good"? So you mean `ConditionHealthCheckFailed`?

File: docs/proposals/20191030-machine-health-checking.md [hunk 10]
```diff
@@ -105,14 +104,17 @@ As an operator of a Management Cluster, I want my machines to be self-healing an
...
+If a health check passes after it has failed, it should remove the ConditionOwnerRemediationSucceeded condition and set ConditionHealthCheckSucceeded to True.
```
[the remark this answers is not in the document]
Comment (benmoss) [reply docs/proposals/20191030-machine-health-checking.md:114]: This is a good point, https://github.com/kubernetes/kubernetes/issues/7856#issuecomment-517240006 brings this up

File: docs/proposals/20191030-machine-health-checking.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (detiber) [reply docs/proposals/20191030-machine-health-checking.md:113]: Indeed, I'm just thinking of the fact that we aren't really able to verify that a Machine is actually Healthy, we are looking to trigger behavior on the case that the health check fails rather than succeeds, and we aren't really trying to block any additional behavior/processes until the health check passes.

File: docs/proposals/20191030-machine-health-checking.md [hunk 11]
```diff
@@ -105,14 +104,17 @@ As an operator of a Management Cluster, I want my machines to be self-healing an
...
+ConditionOwnerRemediationSucceeded is set to False after a health check fails, but should be changed to True by the owning controller after remediation succeeds.
+If remediation fails ConditionOwnerRemediationSucceeded can be updated to a higher severity and the reason can be updated to aid in troubleshooting.
+In the event that ConditionOwnerRemediationSucceeded is still False when the next health check is run, the MHC controller will not update this condition.
```
[the remark this answers is not in the document]
Comment (benmoss) [reply docs/proposals/20191030-machine-health-checking.md:117]: Hmm, this one is stretching my brain a little. When we increase the conditions we increase the number of dimensions of weird state combinations we can get. Not even including the possibility of some conditions being unknown or not present.

h = healthcheck succeeded True
n = remediation needed True
s = remediation succeeded True
H = healthcheck succeeded False
N = remediation needed False
S = remediation succeeded False

with 2 conditions:
- Hs = good, "happy path"
- HS = a failure to remediate, but we can retry or leave for manual intervention
- hs = good, this will be temporarily the state when the deletion of the machine is pending or if/when we support non-deletion remediation
- hS = undefined, if remediation fails but health check is needed it's not clear how we got here. Maybe subsequent healthchecks passed for *reasons*, or the healthcheck was changed? Same concerns as raised in your comment [here](https://github.com/kubernetes-sigs/cluster-api/pull/3056/files#r426709068).

with 3 conditions:
- Hns = good, same as Hs
- HnS = good, same as HS
- HNs = undefined, failing healthcheck, don't need remediation, remediation succeeded
- HNS = undefined, failing healthcheck, don't need remediation, remediation failed
- hns = good, same as hs
- hNs = undefined, passing healthcheck, don't need remediation, remediation succeeded
- hNS = undefined, passing healthcheck, don't need remediation, remediation succeeded
- hnS = undefined, passing healthcheck, need remediation, remediation succeeded

**Comment by vincepri** [comment]
@benmoss What's the status of this proposal? Is there any conversations we can resolve? Let's try to get it merge-able and implementable by EOW if possible

**Comment by benmoss** [comment]
Remaining questions I know of:
- @JoelSpeed @n1r1 and @enxebre want to have a conversation about external remediation and how this fits in [ref](https://github.com/kubernetes-sigs/cluster-api/pull/3056#discussion_r425248169)
- @enxebre feels strongly that the fact that this proposal introduces weird behavior around overlapping MHCs is not acceptable [ref](https://github.com/kubernetes-sigs/cluster-api/pull/3056#discussion_r425754209)
- If the healthcheck passes but remediation has already been failed, what should be done? [ref](https://github.com/kubernetes-sigs/cluster-api/pull/3056#discussion_r426709068)
- Should we go false-y with `ConditionOwnerRemediationNeeded` [ref](https://github.com/kubernetes-sigs/cluster-api/pull/3056#discussion_r426683540)
- Is the a possibility of updates to Conditions clobbering other conditions due to array update/patch semantics? [ref](https://github.com/kubernetes-sigs/cluster-api/pull/3056#discussion_r426687551)

**Comment by vincepri** [comment]
@benmoss Any updates for this proposal after last week's meeting?

File: docs/proposals/20191030-machine-health-checking.md (hunk 8 above)
Comment (vincepri) [reply docs/proposals/20191030-machine-health-checking.md:191]: We're tracking this in https://github.com/kubernetes-sigs/cluster-api/issues/3057

File: docs/proposals/20191030-machine-health-checking.md (hunk 6 above)
Comment (benmoss) [reply docs/proposals/20191030-machine-health-checking.md:173]: We can punt this to the external remediation proposal and never change machines back to healthy for now

**Comment by benmoss** [comment]
We got consensus around moving forward with moving forward on this and leaving the bigger problems around external remediation to that proposal. I'm not sure if @enxebre wants to discuss the problem around overlapping checks more. I think we are going to punt on partial remediation failures by never changing a failing machine back to healthy.

I spoke to @detiber around the update/patch semantics problem, it seems like we're going to have to use `Update` exclusively whenever we change Machine conditions, since we will have two controllers that are both updating and we don't want them to clobber each other's changes.

**Comment by benmoss** [comment]
/hold cancel

If this looks good to people in its current form can we get some lgtms?

**Comment by vincepri** [comment]
Needs squash

File: docs/proposals/20191030-machine-health-checking.md [hunk 12]
```diff
@@ -164,10 +165,10 @@ type target struct {
 
 - Calculate the number of unhealthy targets.
 - Compare current number against `maxUnhealthy` threshold and temporary short circuits remediation if the threshold is met.
-- Annotates unhealthy target machines as described above.
+- Marks unhealthy target machines with the conditions as described above.
 
 Out of band:
-- The owning controller observes the annotation and is responsible to remediate the machine.
+- The owning controller observes the OwnerRemediationSucceeded=False condition and is responsible to remediate the machine.
```
Comment (vincepri) [inline docs/proposals/20191030-machine-health-checking.md:171]:
```suggestion
- The owning controller observes the OwnerRemediated=False condition and is responsible to remediate the machine.
```
